system/memrw16.h

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: GPL-2.0
+#ifndef MEMRW16_H
+#define MEMRW16_H
+/**
+ * \file
+ *
+ * Provides some 16-bit memory access functions. These stop the compiler
+ * optimizing accesses which need to be ordered and atomic. Mostly used
+ * for accessing memory-mapped hardware registers.
+ *
+ *//*
+ * Copyright (C) 2021-2022 Martin Whitaker.
+ */
+
+#include <stdint.h>
+
+/**
+ * Reads and returns the value stored in the 16-bit memory location pointed
+ * to by ptr.
+ */
+static inline uint16_t read16(const volatile uint16_t *ptr)
```

Comment:
I think that we should take this as an opportunity to macro-ify the simple (SISD, temporal) memrw functions as well. Sam, Martin ?

Reply:
Why do you want macros? In-line functions are just as efficient, and far more informative about parameter and return types.

Reply:
Oh, I meant macro-ifying the definitions of the static inline functions themselves :)
The differences between the 8-bit, 16-bit, 32-bit and 64-bit versions are the value type and the width of the assembly instructions. With this patch, the four system/memrw*.h files supporting 3 architectures will amount to nearly 400 lines of quasi-duplicated code and comments; should ARM or RISC-V vendors join the fun, the duplication will worsen.

Reply:
I've just spent time implementing what I had in mind on that front, and pushed the WIP result on the `memrw_refactor` branch. It's slightly less straightforward than I had originally envisioned, because one cannot just put #if chains inside macros, but there are other ways to achieve what I wanted :)

Adding support for another ISA becomes about a dozen lines of code, adding the 8-bit + 16-bit function becomes about 6 lines of code... and the likelihood of bugs, such as the unused `flush64()` currently using `movl`, is reduced, still without slaughtering maintainability, IMO.

Reply:
> I've just spent time implementing what I had in mind on that front, and pushed the WIP result on the `memrw_refactor` branch. It's slightly less straightforward than I had originally envisioned, because one cannot just put #if chains inside macros, but there are other ways to achieve what I wanted :)
> 
> Adding support for another ISA becomes about a dozen lines of code, adding the 8-bit + 16-bit function becomes about 6 lines of code... and the likelihood of bugs, such as the unused `flush64()` currently using `movl`, is reduced, still without slaughtering maintainability, IMO.

Hi @debrouxl ,
The code is very creative, and your ideas is very clever, one more question:
Is the WIP not finished yet? I don't see the 8-bit and 16-bit operations.
![image](https://github.com/user-attachments/assets/36de9239-fccd-465b-8122-f622dbeef2d5)

Reply:
That's correct, it's not finished. I added the LA64 support to show how I envision things to work for multi-ISA support on top of the refactor, but I didn't add the 8-bit and 16-bit operations because they'll be added by this PR.

Should Martin and Sam not frown on my changes, I'd integrate a variant of my commit (switching to a single memrw.h file and removing the LA64 support) to the `main` branch, leaving the addition of LA64 support + the 8-bit and 16-bit operations to you :)

Reply:
I love one-liners and it's for sure creative, so fine to me. 
It's way more understandable that some obscure bit manipulation macros we have. 😅

Reply:
@debrouxl, in general I'm happy with the change, but I'd like to see the Doxygen documentation of the files and functions restored.

Reply:
I've just pushed a `memrw_refactor_v2` branch which contains the single `memrw.h` file I mentioned, and the slight fallout in other files.
The Doxygen documentation of the file was and is there, though it's applied to the definition of `__*_FUNC()` macros only... maybe you mean that the Doxygen comments should be applied to the `__*_FUNC(32/64)` instances ? It doesn't seem to help Doxygen generate documentation for the `read*()` functions, though.

Reply:
The documentation of the `memrw32.h` / `memrw64.h` files is no longer there. This is what I get from your branch:
![screenshot2a](https://github.com/user-attachments/assets/40dfac64-9056-438b-89da-8de9f3b6464e)
compared to this from the main branch:
![screenshot1a](https://github.com/user-attachments/assets/d7e58d50-4453-4030-86fe-04b78dfc09c4)
![screenshot1b](https://github.com/user-attachments/assets/8a83b444-7061-4252-976e-821f07858e45)
Yes, the `memrw.h` file has documentation:
![screenshot2b](https://github.com/user-attachments/assets/1db4cf2c-702e-4be0-bdeb-8aad23f43ff2)
![screenshot2c](https://github.com/user-attachments/assets/aab71e47-e9d7-4015-81d1-efdd7641a3fb)
but I don't think that's as helpful.

You can get Doxygen to generate documentation for the real functions by setting the `EXPAND_AS_DEFINED` variable in the `Doxygen` file to
````
EXPAND_AS_DEFINED      = __READ_FUNC __WRITE_FUNC __FLUSH_FUNC
````

Reply:
Alright. I have modified the commit on the `memrw_refactor_v2` branch to add the change to Doxyfile you suggested, and duplicate the comments to the uses of the macros, which restores documentation for the individual functions :)

Reply:
Thanks. Of course I don't know if anyone actually uses the Doxygen documentation...